File: ios/Releaf/DesignSystem/Components/LeafColorPicker.swift
/*
 * LeafColorPicker.swift
 *
 * Small, reusable color picker keyed off the four leaf theme tokens
 * (coral / green / yellow / dry). Used in the create-shelf and
 * create-notebook flows to let users pick a color *symbolically*
 * (i.e. by leaf-theme name) rather than via a generic hex picker.
 *
 * Each chip is a 32pt circle filled with the theme's primary color,
 * with a thin coral ring around the active selection. Tap to pick.
 * Spacing matches the rest of the app's chip rows.
 *
 * The selected value is the brand's `colorToken` string ("coral",
 * "green", "yellow", "dry") — the same shape the data layer
 * already stores on `Notebook.colorToken` and the Shelf entity.
 */

import SwiftUI

public struct LeafColorPicker: View {
    /// The currently-selected leaf-theme token.
    @Binding public var selection: String
    /// When true, a small `releaf · {token}` eyebrow preview
    /// renders below the chip row in the selected color so the
    /// user can see how their pick will read in the rest of the
    /// chrome before committing. Off by default to keep the
    /// picker compact at call sites that don't need it.
    public let showPreview: Bool

    /// All four leaf themes ordered the way the rest of the app
    /// renders them — coral (default) → green → yellow → dry.
    private let options: [(token: String, label: String, color: Color)] = [
        ("coral",  "Coral",  AppColors.themeCoralPrimary),
        ("green",  "Green",  AppColors.themeGreenPrimary),
        ("yellow", "Yellow", AppColors.themeYellowPrimary),
        ("dry",    "Dry",    AppColors.themeDryPrimary),
    ]

    public init(selection: Binding<String>, showPreview: Bool = false) {
        self._selection = selection
        self.showPreview = showPreview
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: AppSpacing.s2) {
            HStack(spacing: AppSpacing.s3) {
                ForEach(options, id: \.token) { option in
                    Button {
                        selection = option.token
                    } label: {
                        chip(for: option)
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel(Text(option.label))
                    .accessibilityAddTraits(option.token == selection ? [.isSelected] : [])
                }
            }
            if showPreview {
                let palette = ShelfTheme.palette(for: selection)
                HStack(spacing: AppSpacing.s1) {
                    LeafDropletGlyph(tint: palette.background, size: 11)
                    Text("releaf · \(selection)".uppercased())
                        .font(AppText.eyebrow)
                        .tracking(AppLetterSpacing.eyebrow)
                        .foregroundStyle(palette.background)
                }
                .accessibilityElement(children: .combine)
                .accessibilityLabel(Text("Preview · \(selection)"))
            }
        }
    }

    private func chip(for option: (token: String, label: String, color: Color)) -> some View {
        let isActive = option.token == selection
        return ZStack {
            Circle()
                .fill(option.color)
                .frame(width: 32, height: 32)
            // Active chip — thin ring around the chip in the same
            // color, drawn slightly outside so the chip doesn't
            // visually shrink when selected.
            if isActive {
                Circle()
                    .stroke(option.color, lineWidth: 2)
                    .frame(width: 40, height: 40)
                Circle()
                    .stroke(AppColors.cardSolid, lineWidth: 2)
                    .frame(width: 36, height: 36)
            }
        }
        .frame(width: 40, height: 40)
        .contentShape(Circle())
    }
}

#if DEBUG
struct LeafColorPicker_Previews: PreviewProvider {
    struct Host: View {
        @State private var token: String = "green"
        var body: some View {
            VStack(alignment: .leading, spacing: AppSpacing.s3) {
                Text("PICK A COLOR")
                    .font(AppText.eyebrow)
                    .tracking(AppLetterSpacing.eyebrow)
                    .foregroundStyle(AppColors.themeGreenDeep)
                LeafColorPicker(selection: $token)
                Text("Selected: \(token)")
                    .font(AppText.meta)
                    .foregroundStyle(AppColors.textSecondary)
            }
            .padding()
            .background(AppColors.canvas)
        }
    }
    static var previews: some View { Host() }
}
#endif

File: android/app/src/main/java/app/releaf/mobile/ui/components/LeafColorPicker.kt
/*
 * LeafColorPicker.kt
 *
 * Small, reusable color picker keyed off the four leaf-theme
 * tokens (coral / green / yellow / dry). Used in the create-shelf
 * and create-notebook flows so users pick a color *symbolically*
 * by leaf-theme name rather than via a generic hex picker.
 *
 * Each chip is a 32dp circle filled with the theme's primary
 * color, with a thin ring around the active selection. Tap to
 * pick. Mirrors `LeafColorPicker.swift` byte-for-byte on the
 * mapping rule.
 *
 * The selected value is the brand's `colorToken` string ("coral",
 * "green", "yellow", "dry") — same shape as
 * `Notebook.colorToken` and the Shelf entity.
 */

package app.releaf.mobile.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.selection.selectable
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import app.releaf.mobile.ui.theme.AppColors
import app.releaf.mobile.ui.theme.AppSpacing
import app.releaf.mobile.ui.theme.AppTypography

private data class LeafColorOption(
    val token: String,
    val label: String,
    val color: Color,
)

@Composable
private fun leafOptions(): List<LeafColorOption> = listOf(
    LeafColorOption("coral",  "Coral",  AppColors.ThemeCoralPrimary),
    LeafColorOption("green",  "Green",  AppColors.ThemeGreenPrimary),
    LeafColorOption("yellow", "Yellow", AppColors.ThemeYellowPrimary),
    LeafColorOption("dry",    "Dry",    AppColors.ThemeDryPrimary),
)

@Composable
fun LeafColorPicker(
    selection: String,
    onSelect: (String) -> Unit,
    modifier: Modifier = Modifier,
    /** When true, a small `releaf · {token}` eyebrow preview
     *  renders below the chip row in the selected color so the
     *  user can see how their pick will read in the rest of the
     *  chrome before committing. Off by default. */
    showPreview: Boolean = false,
) {
    Column(
        modifier              = modifier,
        verticalArrangement   = Arrangement.spacedBy(AppSpacing.s2),
    ) {
        Row(
            horizontalArrangement = Arrangement.spacedBy(AppSpacing.s3),
            verticalAlignment     = Alignment.CenterVertically,
        ) {
            leafOptions().forEach { option ->
                ColorChip(
                    option   = option,
                    isActive = option.token == selection,
                    onClick  = { onSelect(option.token) },
                )
            }
        }
        if (showPreview) {
            val palette = ShelfTheme.palette(selection)
            Row(
                horizontalArrangement = Arrangement.spacedBy(AppSpacing.s1),
                verticalAlignment     = Alignment.CenterVertically,
            ) {
                LeafDropletGlyph(tint = palette.background, size = 11.dp)
                Text(
                    text  = "releaf · $selection".uppercase(),
                    style = AppTypography.Eyebrow,
                    color = palette.background,
                )
            }
        }
    }
}

@Composable
private fun ColorChip(
    option: LeafColorOption,
    isActive: Boolean,
    onClick: () -> Unit,
) {
    // `Modifier.selectable` provides correct accessibility (RadioButton role
    // + selected state) and the click handling in one call.
    Box(
        modifier = Modifier
            .size(40.dp)
            .selectable(
                selected = isActive,
                role     = Role.RadioButton,
                onClick  = onClick,
            ),
        contentAlignment = Alignment.Center,
    ) {
        if (isActive) {
            // Outer ring in the chip's own colour.
            Box(
                modifier = Modifier
                    .size(40.dp)
                    .clip(CircleShape)
                    .border(2.dp, option.color, CircleShape),
            )
            // Cream insert separates the outer ring from the chip
            // so the active state reads as a halo, not a thicker
            // chip.
            Box(
                modifier = Modifier
                    .size(36.dp)
                    .clip(CircleShape)
                    .border(2.dp, AppColors.CardSolid, CircleShape),
            )
        }
        Box(
            modifier = Modifier
                .size(32.dp)
                .clip(CircleShape)
                .background(option.color),
        )
    }
}

@Preview(showBackground = true, backgroundColor = 0xFFF5EEDF, widthDp = 320)
@Composable
private fun LeafColorPickerPreview() {
    val state = androidx.compose.runtime.remember { androidx.compose.runtime.mutableStateOf("green") }
    Box(
        modifier = Modifier
            .background(AppColors.Canvas)
            .padding(AppSpacing.s4),
    ) {
        LeafColorPicker(
            selection = state.value,
            onSelect  = { state.value = it },
        )
    }
}
